### ISSUE TITLE
Order of Application Gateway Configuration Blocks Should be Irrelvant

### ISSUE THREAD

**Issue by schwarzzz** [hashicorp/terraform-provider-azurerm#6896]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform (and AzureRM Provider) Version

Terraform v0.12.24
+ provider.azurerm v2.8.0

### Affected Resource(s)

* `azurerm_application_gateway`

### Issue Description

The order of certain configuration blocks (like `http_listener`) seems to relevant to the azurerm provider. If you swap the order of two blocks in your configuration file, the azurerm provider plans to update both existing elements (by replacing each attribute). For other blocks (like `probe`) this is working fine.

### Terraform Configuration Files

```hcl
# Copy-paste your Terraform configurations here - for large Terraform configs,
# please use a service like Dropbox and share a link to the ZIP file. For
# security, you can also encrypt the files using our GPG public key: https://keybase.io/hashicorp
```

### Expected Behavior

The order of the configuration blocks should be irrelevant. Terraform should not plan any updates, if only the order of two blocks changes.

### Actual Behavior

Terraform sticks to the order in the configuration file and plans unnecessary updates.

### Steps to Reproduce

Define an `azurerm_application_gateway` resource with at least 2 configuration blocks of the following types:
+ `http_listener`
+ `backend_http_settings`
+ `redirect_configuration`
+ `ssl_certificate`
+ `request_routing_rule`
+ `probe`

Deploy the application gateway.

Swap the order of each configuration block pair.

Plan a new deployment and investigate the plan:

While there is no change in the list `request_routing_rule` and `probe`, all other blocks are marked for modification. Terraform plans to swap the content of each element.

**Comment by MaxiPalle** [comment]
Experienced this too.

**Comment by neil-yechenwei** [comment]
@schwarzzz , thanks for opening this issue. I just want to double confirm you mean there is diff after swapped the order of `http_listener` and `backend_http_settings`, right? Or you mean there is diff after swapped the multiple blocks of `http_listener`. If it's possible, could you share your tfconfig and error message? Thanks.

**Comment by mazilu88** [comment]
we have a module with multiple dynamic blocks. when adding rules the output gets really crazy saying it will delete and recreate the same block. on other resources just updates in place with different names

`  dynamic "request_routing_rule" {
    for_each = var.request_routing_rules_http
    content {
      name                        = request_routing_rule.value.name
      rule_type                   = request_routing_rule.value.is_path_based ? "PathBasedRouting" : "Basic"
      http_listener_name          = request_routing_rule.value.http_listener_name
      redirect_configuration_name = request_routing_rule.value.redirect_configuration_name
      url_path_map_name           = request_routing_rule.value.url_path_map_name
    }
  }

  dynamic "request_routing_rule" {
    for_each = var.request_routing_rules_https
    content {
      name                       = request_routing_rule.value.name
      rule_type                  = request_routing_rule.value.is_path_based ? "PathBasedRouting" : "Basic"
      http_listener_name         = request_routing_rule.value.http_listener_name
      backend_address_pool_name  = request_routing_rule.value.backend_address_pool_name
      backend_http_settings_name = request_routing_rule.value.backend_http_settings_name
      url_path_map_name          = request_routing_rule.value.url_path_map_name
      rewrite_rule_set_name      = "xforwarded-realip-set"
    }
  }`

`          - request_routing_rule {
          - backend_address_pool_id    = "providers/Microsoft.Network/applicationGateways/sameday-app-gw-prod/backendAddressPools/lm-api-prod" -> null
          - backend_address_pool_name  = "lm-api-prod" -> null
          - backend_http_settings_id   = providers/Microsoft.Network/applicationGateways/sameday-app-gw-prod/backendHttpSettingsCollection/lm-api-sett-80" -> null
          - backend_http_settings_name = "lm-api-sett-80" -> null
          - http_listener_id           = providers/Microsoft.Network/applicationGateways/sameday-app-gw-prod/httpListeners/lm-api-443-listen" -> null
          - http_listener_name         = "lm-api-443-listen" -> null
          - id                         = "//providers/Microsoft.Network/applicationGateways/sameday-app-gw-prod/requestRoutingRules/lm-api-443-rule" -> null
          - name                       = "lm-api-443-rule" -> null
          - rewrite_rule_set_id        = "providers/Microsoft.Network/applicationGateways/sameday-app-gw-prod/rewriteRuleSets/xforwarded-realip-set" -> null
          - rewrite_rule_set_name      = "xforwarded-realip-set" -> null
          - rule_type                  = "Basic" -> null
        }
      + request_routing_rule {
          + backend_address_pool_id    = "providers/Microsoft.Network/applicationGateways/sameday-app-gw-prod/backendAddressPools/lm-api-prod"
          + backend_address_pool_name  = "lm-api-prod"
          + backend_http_settings_id   = "providers/Microsoft.Network/applicationGateways/sameday-app-gw-prod/backendHttpSettingsCollection/lm-api-sett-80"
          + backend_http_settings_name = "lm-api-sett-80"
          + http_listener_id           = "providers/Microsoft.Network/applicationGateways/sameday-app-gw-prod/httpListeners/lm-api-443-listen"
          + http_listener_name         = "lm-api-443-listen"
          + id                         = "providers/Microsoft.Network/applicationGateways/sameday-app-gw-prod/requestRoutingRules/lm-api-443-rule"
          + name                       = "lm-api-443-rule"
          + rewrite_rule_set_id        = providers/Microsoft.Network/applicationGateways/sameday-app-gw-prod/rewriteRuleSets/xforwarded-realip-set"
          + rewrite_rule_set_name      = "xforwarded-realip-set"
          + rule_type                  = "Basic"
        }`

**Comment by schwarzzz** [comment]
Thats exactly our scenrario!
But you'd also run into trouble with only one dynamic block of a certain type...

In our code, we iterate over a list of complex object and build up a gateway configuration with multiple listeners/routing rules/etc.
Appending new entries to that "input list" is no problem. But removing an entry at position `x` in a list of `n` elements would be bad: Instead of just deleting the element at `x`, all elements from `x` to `n-1` will be updated (with the data of the subsequent element) and finally element `n` will be deleted.

I never tried it, but I'm pretty sure, that such a modification of the gateway would cause at least hickups in production.

**Comment by ArcturusZhang** [comment]
Should be a follow up of #6452

**Comment by Nyxbiker** [comment]
In my case even after a terraform apply, terraform keeps trying to destroy and re-add (in my case request_routing rules) blocks, only a new application gateway deployment stops it but if you also change something on the new one, it starts all over again.

**Comment by smits23** [comment]
I have a similar scenario, 
Whenever I am adding a new block of routing rule in my configuration, the terraform plan displays inplace delete existing routing rules and create all of them again. Is this going to have an impact in real time for multi tenant scenario?

**Comment by smits23** [comment]
Can I please be notified when this Bugis resolved, it will be a showstopper for our infra deployment. Thank you

**Comment by smits23** [comment]
I also want to add that, I tested by adding two non dynamic blocks of request_routing_rule  to tested this use case, 
test case 1 - Deploy 2 request routing rule  (not using for_each)
test case 2 - Delete one of the routing rule and deploy again and verify the TF plan. 
Result : TF plan shows in place delete and re-creation of all request routing rules

**Comment by norakennedy** [comment]
Looking at the source provider, it seems like the problem is with the application gateway code.  Since indexing matters for routing rules, this needs to be set to TypeList instead of TypeSet: 
![image](https://user-images.githubusercontent.com/39530048/98958903-61ee5300-24c8-11eb-900f-7330f949e3ae.png)

https://github.com/terraform-providers/terraform-provider-azurerm/blob/0c2b8653c50ec16487376cd9d546d137a5f9ae9a/azurerm/internal/services/network/application_gateway_resource.go#L475

Same with this one:
![image](https://user-images.githubusercontent.com/39530048/98959045-864a2f80-24c8-11eb-977c-45601303c4eb.png)

https://github.com/terraform-providers/terraform-provider-azurerm/blob/0c2b8653c50ec16487376cd9d546d137a5f9ae9a/azurerm/internal/services/network/application_gateway_resource.go#L565

Any ETA on a fix? @simonbrady @kvendingoldo @jackofallops

**Comment by mazilu88** [comment]
hello guys, you need to follow the bread crumbs aka  "Should be a follow up of #6452" which leads to "this should be mentioned in #6896 and possibly fixed by #7021"

#7021  is the key which states that it will be released in azurerm 3.0

**Comment by w3bward** [comment]
This is one of the biggest annoyances when making changes to app gateways. Most of the changes I make to a gateway is the addition of a listener/rule combination, which are handled dynamically in my config based on variables because I have to deploy to separate environments. Adding or removing a single item causes terraform to plan the deletion and addition of hundreds of lines of configuration; the plans produced are basically useless for reviewing what changes are actually about to be applied.

I see in #7021 that the fix was submitted in a PR, but it was been deemed a breaking change and pushed off to v3.0 of the provider. I'm unclear as to why the fix would be a breaking change for existing configs in most use cases. Wouldn't the actual config syntax for the change from schema.TypeList to schema.TypeSet be basically identical except that TypeSet doesn't allow for duplicates? I suppose it's possible that someone has the same config block with exactly the same values defined twice on an app gateway, but that seems like an edge case and most likely accidental as such a configuration wouldn't provide any functional value (would such a configuration even be accepted by the api?). If this change is breaking, isn't that what the ability to specify provider version requirements is for?

Breaking changes have been introduced between major versions previously. It's really disappointing be stuck waiting for a major revision that has been under development for over a year and has no ETA just to get a simple fix that has already been developed and solves a major headache. 

TL;DR, I feel like this change should be prioritized for release before v3.0 of the provider.

**Comment by jhyot** [comment]
This is also making Application Gateways difficult to manage via Terraform for us, and we would appreciate a fix in an upcoming minor version, or transitioning to the next major version soon.

**Comment by prubis** [comment]
@mbfrahry not sure this has been fixed with the new release, check out the re-arranging TF is proposing to do to my http_listeners below:

Version:
```
prubis:/home/prubis/repos/web$ tf -version
Terraform v1.1.6
on linux_amd64
+ provider registry.terraform.io/hashicorp/azuread v2.18.0
+ provider registry.terraform.io/hashicorp/azurerm v2.97.0
```

Code:
```hcl
  dynamic "http_listener" {
    for_each = var.app_gateway_web_configs

    content {
      name                           = "${http_listener.key}-https"
      frontend_ip_configuration_name = "FrontendIp"
      frontend_port_name             = "port_${http_listener.value.listener_port}"
      host_name                      = http_listener.value.listener_host_name
      protocol                       = "Https"
      require_sni                    = true
      ssl_certificate_name           = http_listener.value.ssl_certificate_name
    }
  }
```

Variables:
```hcl
app_gateway_web_configs = {
  web = {
    listener_host_name   = "bi.aaa.com"
    listener_port        = 443
    backend_host_name    = "bi.aaa.com"
    backend_ips          = ["1.2.3.4"]
    backend_port         = 443
    ssl_certificate_name = "bi-aaa-com"
  }
  sync = {
    listener_host_name   = "bisync.aaa.com"
    listener_port        = 443
    backend_host_name    = "bisync.aaa.com"
    backend_ips          = ["1.2.3.5"]
    backend_port         = 443
    ssl_certificate_name = "bi-aaa-com"
  }
  upload = {
    listener_host_name   = "biupload.aaa.com"
    listener_port        = 443
    backend_host_name    = "vmname.aaa.com"
    backend_ips          = ["1.2.3.4"]
    backend_port         = 443
    ssl_certificate_name = "bi-aaa-com"
  }
  partner = {
    listener_host_name   = "partner.bi.aaa.com"
    listener_port        = 443
    backend_host_name    = "partner.bi.aaa.com"
    backend_ips          = ["1.2.3.6"]
    backend_port         = 443
    ssl_certificate_name = "bi-aaa-com"
  }
  ddp = {
    listener_host_name   = "bi.aaa.com"
    listener_port        = 8443
    backend_host_name    = "bi.aaa.com"
    backend_ips          = ["1.2.3.4"]
    backend_port         = 8443
    ssl_certificate_name = "bi-aaa-com"
  }
}
```

And then the proposal of `terraform apply`:
```
  # azurerm_application_gateway.main will be updated in-place
  ~ resource "azurerm_application_gateway" "main" {
  [...]

      ~ http_listener {
          ~ host_name                      = "bisync.aaa.com" -> "partner.bi.aaa.com"
            id                             = "/subscriptions/.../applicationGateways/appgw/httpListeners/sync-https"
          ~ name                           = "sync-https" -> "partner-https"
            # (9 unchanged attributes hidden)
        }
      ~ http_listener {
          ~ host_name                      = "biupload.aaa.com" -> "bisync.aaa.com"
            id                             = "/subscriptions/..../applicationGateways/appgw/httpListeners/upload-https"
          ~ name                           = "upload-https" -> "sync-https"
            # (9 unchanged attributes hidden)
        }
      ~ http_listener {
          ~ host_name                      = "bi.aaa.com" -> "biupload.aaa.com"
            id                             = "/subscriptions/..../applicationGateways/appgw/httpListeners/web-https"
          ~ name                           = "web-https" -> "upload-https"
            # (9 unchanged attributes hidden)
        }
      ~ http_listener {
          ~ host_name                      = "partner.bi.aaa.com" -> "bi.aaa.com"
            id                             = "/subscriptions/..../applicationGateways/appgw/httpListeners/partner-https"
          ~ name                           = "partner-https" -> "web-https"
            # (9 unchanged attributes hidden)
        }

     # (26 unchanged blocks hidden)
}
```

**Comment by mbfrahry** [comment]
Hey @prubis, I can take a look at this tomorrow but do you mind confirming that you're using the 3.0 beta? It's enabled by setting `ARM_THREEPOINTZERO_BETA` to `true` through an environment variable

**Comment by prubis** [comment]
Right, okay, that I wasn't. I'll give it a shot and report back.

**Comment by prubis** [comment]
Hi @mbfrahry, okay, yes you're right it was the 3.0 beta not being enabled. I've re-run tf plan with the env variable set and now it's not planning on re-arranging them. This is in a production environment so I can't actually run the apply, but once it's moved out of beta I can expect it to work. Thanks!

**Comment by mbfrahry** [comment]
That's great news! Thanks for confirming @prubis

**Comment by johannespetereit** [comment]
sorry, but since most comments in #16136 seem to be marked as "off topic", could you please reopen this original issue. If you read the original post in this issue carefully, it doesn't really matter what you change internally, if the end result is that **the issue still persists in 3.x.x**. it's nice that some internal restructuring took place, but right now the state is "we did all we could, but terraform reports a change, because the azure api returns unordered lists" (as is my understanding, please correct me). This should not be a reason to close an issue as solved.

As an architect, this is pointing to a (really consistant and de facto industry standard) **upstream** APIs as the scape goat. If you go further down the ticket rabbithole, you even reach the point where Azure itself is the problem, not sorting the returned values for the terraform provider pass through to terraform without additional transformations. Which is frustrating, because my customers think I'm insane pointing out that an MS framework is buggy because of another MS framework, and all the tickets get put off because of "not our problem".

In my experience you shouldn't rely on upstream apis to change, if you want to deliver a successful product.  Sorry for the rant,  but trying to get a good customer experience is really hard, if all you get as support is a 🤷‍♀️ with pointing to other (MS) teams. This is especially hard to sell, if those teams all work for the same company (Microsoft). 

While the benefits of open sourcing everything are really overwhelmingly positive (kudos on that note!), it is frustrating to realize how fragmented the development teams have become.  This critisism of course exempts all the hard working contributors out there that are pushing their **free time** into a product for pure love, that a company is later advertising for in their own name, and indirectly using to up cloud sales. 

---

**tl;dr.** terraform is in my view currently the industry leader in IaC solution for clouds. MS to this day fails to properly support provisioning their only (and admittedly really great) reverse proxy solution in terraform. I do not have one customer not using AGW, so this makes this a blocking issue for us in multiple contexts, with many coleagues of mine being fed up with 1000s of lines of terraform changes for any go live. I will therefore advice building back terraform in favor of az / PowerShell regarding AGW, as this is a much more fool prove way to update it (contact me at forename.name@adesso.de and I will provide samples. This is not for privacy reasons, but the size of the scripts).

Forgive my frustration, but my collegues and I were looking forward to a fix for multiple years, and communicating the progress (documented above) to our customers, only to realize after days of migrations that this will not happen in the forseeable future, because "terraform or the Azure API will have to change". :/

**Comment by opslivia** [comment]
hi @johannespetereit, I'm on the AzureRM team and I'd like to talk with you more about this. I tried emailing you at the provided address (using your name as stated on your profile), but it keeps bouncing back. Would you like to [sign up for a one-on-one](https://calendly.com/hashicorp-oliviac/tf-azure-providers-experience?month=2022-04) with me?

**Comment by johannespetereit** [comment]
@opslivia Thanks! looking forward to it. 

Could it be viable to implement a workaround [like proposed in by murdibb in #10388](https://github.com/hashicorp/terraform-provider-azurerm/issues/10388#issuecomment-1067095863)? Or would that be in vain because the order of the blocks in the datasource / resource are indeterministic?
